### ISSUE TITLE
as.ctd(argo, profile=...) should extract relevant fields from time, lon, lat, etc

### ISSUE THREAD

**Issue by richardsc** [dankelley/oce#1176]
I noticed that the recent `as.ctd(argo, profile...)` functionality (which I have been using lots, BTW) extracts on the profile from the *matrix* components, but not from the *vector* fields in the `@data` slot (like `time`, `longitude`, etc). E.g.
```r
library(oce)
data(argo)
str(as.ctd(argo, profile=1)@data)
```
```
List of 12
 $ time                    : POSIXct[1:207], format: "2005-10-29 13:52:11" "2005-11-08 13:52:13" "2005-11-18 13:46:20" "2005-11-28 14:02:56" ...
 $ latitude                : num [1:207] 61 60.8 61.1 61.3 61.5 ...
 $ longitude               : num [1:207] -21.4 -21.9 -22.3 -22.4 -23.1 ...
 $ pressure                : num [1:56] 4.8 9.2 14.6 19.3 24 ...
 $ pressureAdjusted        : num [1:56] 4.8 9.2 14.6 19.3 24 ...
 $ pressureAdjustedError   : num [1:56] 2.4 2.4 2.4 2.4 2.4 ...
 $ salinity                : num [1:56] 35.2 35.2 35.2 35.2 35.2 ...
 $ salinityAdjusted        : num [1:56] 35.2 35.2 35.2 35.2 35.2 ...
 $ salinityAdjustedError   : num [1:56] 0.02 0.02 0.02 0.02 0.02 ...
 $ temperature             : num [1:56] 9.71 9.71 9.71 9.71 9.71 ...
 $ temperatureAdjusted     : num [1:56] 9.71 9.71 9.71 9.71 9.71 ...
 $ temperatureAdjustedError: num [1:56] 0.002 0.002 0.002 0.002 0.002 ...
```

I think that it *should* extract such (single) values, but also it should probably move them to the `@metadata` slot to be more consistent with a "typical" CTD profile? That would mean that `time`  would become `@metadata$startTime`, etc.

Thoughts @dankelley ?

**Comment by dankelley** [comment]
Yes, very good call!! I think it should take just the first, and it should do as you say. I guess it might be smart to also look for other metadata that could get filled in. For example, the "station" could be filled in as the profile number (if that's not done already).

**Comment by richardsc** [comment]
Oh yes -- good idea on the `station` metadata

**Comment by richardsc** [comment]
I'm just having a look at the code, and wondering if it needs a bit of an overhaul. Currently the code checks each field as to whether or not it is a matrix, and then if the `profile` arg is given it will pull out the relevant profile. However, that doesn't work for `time`, 'longitude`, etc, giving rise to the current issue. 

I wonder if a better approach is a "branching" higher up to test for an `argo` object -- that way the specific conversion can be handled in it's own consistent region of code and not bundled in with something trying to be more generic (are there other instances of ctd data that is stored in matrices?).

The relevant section of code is:
```r
        convertedMatrix <- FALSE
        for (field in names(d)) {
            if (field == "time") {
                if (length(d$time) > 1) res@data$time <- d$time else res@metadata$time <- d$time
            } else {
                ##res <- ctdAddColumn(res, column=d[[field]], name=field, label=field, log=FALSE)
                dataInField <- d[[field]]
                if (is.matrix(dataInField)) {
                    ## Note that the tests are repeated for each datum. That may not
                    ## be required, but this scheme will be kept unless/until we
                    ## learn that the data are required to have the same dimensionality.
                    if (missing(profile)) {
                        profile <- 1
                        warning("using just column 1 of matrix data; use the 'profile' argument to select a specific profile or try as.section() to keep all columns")
                    }
                    if (!is.numeric(profile) || length(profile) != 1 || profile < 1) {
                        stop("profile must be a positive integer")
                    }
                    ncol <- ncol(d[[field]])
                    if (profile > ncol)
                        stop("profile cannot exceed ", ncol, " for a data matrix with ", ncol, " columns")
                    convertedMatrix <- TRUE
                    res@data[[field]] <- d[[field]][, profile]
                } else {
                    res@data[[field]] <- d[[field]]
                }
            }
        }
```

**Comment by dankelley** [comment]
Good point. I had also thought about the branching business but that was more in the context of efficiency, and I decided not to change the code because looping (or testing for matrix-ness) a thousand times will cost essentially zero time.  I think that if a change is desired, it should be based on the clarity of the code, i.e. a balance between being DRY and being clear. Looking at the code snippet above, it seems to me the solution is, as you propose, to do `isMatrix <- is.matrix(d[[1]])` or something like that, at the start. Also, then we can do tests on whether `col` is too large before doing anything else, which again clarifies.

But before doing this, I wonder about the comment I wrote in the code, about not knowing whether all data will be in matrix form, if any data are in that form. The main reason for the present coding was of uncertainty on this matter. It would be a shame to rewrite this code and have to revert it later, to fix a user-identified bug.  (I never worry that much about bugs that developers find, because any developer has the skill to fix the bugs without help, so usually 1/2 hour will fix the problem and the developer can get back to their paper/proposal/thesis/etc... it's way harder for a user who is unlikely to understand the code layout.)
